reqs/requirements.in

```diff
@@ -1,5 +1,5 @@
-Django>=1.11
-cryptography
-pyjwt
-requests
-requests-oauthlib
+Django>=2.2
+cryptography~=3.0
+pyjwt~=2.0
+requests~=2.24
+requests-oauthlib~=1.3

```

Comment:
I would rather not use `~=` or `<` in the `requirements.in`. Unless there is a known incompatibility and in that case I would rather use a `<` to make sure we do not go to that version. I know I do not pay as much attention to the repo as I should, but I would rather something break and force a downstream application to put a hard `<` limiter to prevent a newer version then to completely break future installations for packages we do not know if they will be incompatible. i.e., this package _may_ be 100% compatible with Django 4.0 when it comes out. We do not know if it will be or not. But if we put `django<4` in the `requirements.in`/`setup.py`, it will forcibly _prevent_ people from installing it on Django 4.0 regardless of it working or not.

The issue with `~=` is that it does not work the same with all libraries. Like `requests` has a stable interface if `requests` ever got a 2.25 version it would be backwards compatible with 2.24 (unlikely since `xhttp` is now the replacement for `requests`). Django use to have a problem with minor versions breaking things, but since they started bumping the major, it has not been an issue any more.

Also, it might be a good idea to remove `cryptography` completely and try to use `pyjwt[crypto]` since I do not think I am using any `cryptography` specific code that is not used by `pyjwt`.

Reply:
I agree that the Django dependency should be as permissive as possible. However, new major version of `pyjwt`, `requests`, or `requests-oauthlib` might break something as those are used more directly by this library and at least `pyjwt` and `requests` follow semantic versioning. I would be more defensive with these dependencies and adjust the upper limit after new major version has been released and verified to be compatible, but I understand this approach would also have its downsides.

I also noticed that `pip 20.3` is much more strict with conflicting dependencies than its earlier versions and the minimum version from these `~=` requirements might cause problems as well.

Reply:
`requests` and `requests-oauthlib` will likely never get updated. `requests` is basically dead in the water and I do not think v3 is ever going to be made. I am pretty sure the only updates it will ever get is cert/security patches. I would rather leave them both as is and just get them removed in favor of `httpx`. 

`pyjwt` it probably would make more sense to add a `<` to to prevent major breaking changes. I would say dig into `pyjwt`'s development cycle a bit and see if they follow strict semantic version, if they do, just do `<3`, if they do not, probably just do `<2.x` where `x` is the latest minor version.